### ISSUE TITLE
training：high CPU usage，the first epoch dies

### ISSUE THREAD

**Issue by yskjAvery** [ultralytics/ultralytics#11749]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Multi-GPU

### Bug

Large amount of data and input size lead to high CPU usage during training, low idle rate, and failure to start training; the first epoch dies.I believe this is a bug.
example：data volume ：55246，imgsz=1088，GPUs are four NVIDIA V100 cards
I have tried two approaches: using a smaller dataset and smaller input size, and both worked fine for training.

The complete error message is as follows：picture
![训练报错](https://github.com/ultralytics/ultralytics/assets/151350131/35de393a-b956-46cb-b54f-2883b49993a0)

### Environment

GPUs are four NVIDIA V100 cards

### Additional

![训练报错](https://github.com/ultralytics/ultralytics/assets/151350131/54b8e77b-532a-40c9-afba-fda87c1d50f3)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by yskjAvery** [comment]
By using the top command to check the CPU idle rate, the idle rate drops to 10%, causing the program to crash. Can we move the data reading to the GPU instead?

**Comment by yskjAvery** [comment]
> Let us know how it goes!
Thank you for your reply. However, the methods you mentioned did not work for me when I tried them today. In order to solve this problem, I attempted several methods that all failed: 1. I tried setting the worker to a larger value of 16 and a smaller value of 2. After loading data for a while in the first epoch, the CPU's idle rate suddenly dropped significantly, below 40% (I monitored the CPU usage throughout using the top command). Then the program crashed. 2. I tried training without using caching. With this method, the program crashed every time at the 4th epoch and the training process was very slow. I have trained custom datasets using YOLOv8 successfully many times in the past, and in those cases, the CPU idle rate was always above 40%. Additionally, I was able to successfully train the dataset that failed by modifying the input image size from the original 1088*1088 to 640*640. However, this is not the experimental result I wanted. I hope you can provide some assistance.

**Comment by yskjAvery** [comment]
![报错](https://github.com/ultralytics/ultralytics/assets/151350131/d0b4e0b3-42df-406c-ae48-4bce7571ee77)

**Comment by glenn-jocher** [comment]
Hello! Thank you for providing the additional details and for your thorough testing. It seems like the high resolution of the images combined with the large dataset is stressing the CPU during data loading, leading to crashes.

One possible approach could be to experiment with reducing the image size dynamically during training. This can significantly alleviate the load on the CPU. If reducing the `imgsz` to 640 works without crashes, you might try increasing it incrementally to find the maximum manageable size.

Another strategy might be to explicitly manage the PyTorch DataLoader's worker initialization function to ensure each worker has a separate seed and potentially better memory management. Here’s a snippet on how you can implement this in your training script:

```python
from torch.utils.data import DataLoader
from torch.utils.data.distributed import DistributedSampler

def worker_init_fn(worker_id):
    np.random.seed(np.random.get_state()[1][0] + worker_id)

# When creating DataLoader
train_loader = DataLoader(
    dataset,
    batch_size=batch_size,
    shuffle=False,
    sampler=DistributedSampler(dataset) if is_distributed else None,
    num_workers=num_workers,
    pin_memory=True,
    worker_init_fn=worker_init_fn,
)
```

If these adjustments do not solve the issue, it might be helpful to look at system logs or increase virtual memory limits as sometimes memory paging issues could cause such anomalies.

Hope this helps! Keep us posted on your progress. 😊

**Comment by yskjAvery** [comment]
> Hope this helps! Keep us posted on your progress.
Thank you for your reply!
       I understand what you mean.I'm sorry, I'm not very familiar with the method you mentioned. The way I train is implemented through the Python method mentioned in the readme document, by creating a task.py file and running task.py for training. The code is as shown in the figure below. How should I use the guidance code you provided in my task.py file? Could you please provide a complete implementation code example? Thank you!
![代码](https://github.com/ultralytics/ultralytics/assets/151350131/1f880211-e98c-4f54-befa-886a88c091d8)

**Comment by glenn-jocher** [comment]
Hello! Thanks for following up. I'm glad to assist!

To integrate the worker initialization into your existing `task.py`, you would modify your DataLoader setup. Below is an enhanced version of your DataLoader setup where `worker_init_fn` is used to manage the random seeds and potentially improve the CPU usage:

```python
import numpy as np
import torch
from torch.utils.data import DataLoader

def worker_init_fn(worker_id):
    np.random.seed(np.random.get_state()[1][0] + worker_id)

# Assuming you're creating a DataLoader somewhere in your task.py
train_loader = DataLoader(
    your_dataset,
    batch_size=your_batch_size,
    shuffle=True,  # or False if using DistributedSampler
    num_workers=your_num_workers,
    pin_memory=True,
    worker_init_fn=worker_init_fn
)
```
Replace `your_dataset`, `your_batch_size`, and `your_num_workers` with your actual dataset and parameters.

This snippet ensures that each worker thread of the DataLoader has a different seed, which might help in managing resources more effectively.

Give it a try and let me know how it goes! 😊

**Comment by yskjAvery** [comment]
> ```python
> train_loader
> ```

Thank you for your patient responses. I am currently confused about two points: 
1. Do I need to write a Dataset class myself for this dataset, then instantiate it to get the dataset? How can I implement this dataset? Can I directly replace it in the original YOLOv8 code, or should I keep it separate? 
2. How can I add this instantiated train_loader to my training code? Currently, if I don't modify the DataLoader in the task.py script, the training code only requires two lines: model = YOLO(), result = model.train(). Based on these two lines of code, what should I do to start training?

**Comment by yskjAvery** [comment]
> > ```python
> > train_loader
> > ```
> 
> Thank you for your patient responses. I am currently confused about two points:
> 
> 1. Do I need to write a Dataset class myself for this dataset, then instantiate it to get the dataset? How can I implement this dataset? Can I directly replace it in the original YOLOv8 code, or should I keep it separate?
> 2. How can I add this instantiated train_loader to my training code? Currently, if I don't modify the DataLoader in the task.py script, the training code only requires two lines: model = YOLO(), result = model.train(). Based on these two lines of code, what should I do to start training?

My current challenge is how to implement the ideas you mentioned in a custom task.py file to complete my training task. In the existing task.py file, only the code model = YOLO(), result = model.train() is needed to complete the training. How can I incorporate the dataset and instantiated train_loader into the training code as you suggested?

**Comment by glenn-jocher** [comment]
Hello! Great questions, let’s clarify them:

1. **Dataset Class**: If you have a custom dataset, you usually need a Custom Dataset Class. However, for training YOLOv8 models, if your dataset follows the expected format (like COCO), you might not need to. Just point the `data` argument in your `model.train()` to the correct configuration YAML that describes your dataset paths and classes.

2. **Integrating DataLoader**: The default `model.train()` method handles DataLoader creation internally based on your dataset configuration YAML file. If you want custom behavior not supported default training method, you would then need to modify the source code of YOLOv8, which might be a bit complex and such changes are not typically done unless necessary.

Here's how you point to a dataset in `model.train()`:

```python
model = YOLO('config.yaml')  # Load model with a specific configuration file
result = model.train()  # Train the model
```

For the `config.yaml`, ensure it refers to where your dataset and class labels reside. Hope this helps clarify your concerns! 😊

**Comment by yskjAvery** [comment]
> Hello! Great questions, let’s clarify them:
> 
> 1. **Dataset Class**: If you have a custom dataset, you usually need a Custom Dataset Class. However, for training YOLOv8 models, if your dataset follows the expected format (like COCO), you might not need to. Just point the `data` argument in your `model.train()` to the correct configuration YAML that describes your dataset paths and classes.
> 2. **Integrating DataLoader**: The default `model.train()` method handles DataLoader creation internally based on your dataset configuration YAML file. If you want custom behavior not supported default training method, you would then need to modify the source code of YOLOv8, which might be a bit complex and such changes are not typically done unless necessary.
> 
> Here's how you point to a dataset in `model.train()`:
> 
> ```python
> model = YOLO('config.yaml')  # Load model with a specific configuration file
> result = model.train()  # Train the model
> ```
> 
> For the `config.yaml`, ensure it refers to where your dataset and class labels reside. Hope this helps clarify your concerns! 😊



> Hello! Great questions, let’s clarify them:
> 
> 1. **Dataset Class**: If you have a custom dataset, you usually need a Custom Dataset Class. However, for training YOLOv8 models, if your dataset follows the expected format (like COCO), you might not need to. Just point the `data` argument in your `model.train()` to the correct configuration YAML that describes your dataset paths and classes.
> 2. **Integrating DataLoader**: The default `model.train()` method handles DataLoader creation internally based on your dataset configuration YAML file. If you want custom behavior not supported default training method, you would then need to modify the source code of YOLOv8, which might be a bit complex and such changes are not typically done unless necessary.
> 
> Here's how you point to a dataset in `model.train()`:
> 
> ```python
> model = YOLO('config.yaml')  # Load model with a specific configuration file
> result = model.train()  # Train the model
> ```
> 
> For the `config.yaml`, ensure it refers to where your dataset and class labels reside. Hope this helps clarify your concerns! 😊

Dear author, hello! The training method you mentioned is the one I am currently using, and all my training parameters are also set in the default.yaml file. I need to reiterate my issue - when using a custom dataset, during training, after the data caching is complete, before the start of the first epoch, the program throws an error and cannot complete the training (my dataset is large with 55,000 images, each with a large input size of 1088*1088). I found that the high CPU usage during data caching leads to a low CPU idle rate, dropping below 40% and causing the program to fail. Your previous strategy involved explicitly managing the worker initialization function of PyTorch DataLoader to ensure that each worker has independent seeds and better memory management. My current need is how to implement your approach in code. Could you provide me with a complete and usable training code?
